### ISSUE TITLE
Circular dependency when using inside $http interceptor

### ISSUE THREAD

**Issue by suhrab** [rubenv/angular-gettext#138]
This article http://blog.altoros.com/speed-up-i18n-in-angularjs.html
describes how to speed up angular applications that heavily use translate libraries like this one. Article gives example for angular-translate. I wanted it work with _angular-gettext_.

In example they use $translate service from angular-translate as dependency to factory that creates request interceptor. It works.

I replaced $translate with _gettextCatalog_, but now I get an error from angular about circular dependency: https://docs.angularjs.org/error/$injector/cdep?p0=$http%20%3C-%20gettextCatalog%20%3C-%20translateHttpInterceptor%20%3C-%20$http%20%3C-%20$templateFactory%20%3C-%20$view%20%3C-%20$state

The problem:
angular-gettext factory _'gettextCatalog'_ depends on $http (loadRemote method). if I remove this dependency and 'loadRemote' method, everything works fine.

_Actually I think that angular-gettext should not depend on $http_, it is very easy to write my own loading from remote method using $http and 'gettextCatalog.setStrings'.

_angular-translate_ does not depend on $http

**Comment by frederikprijck** [comment]
Any updates on this ? Still open, and the issue still exists.
loadRemote should realy be removed from the catalog, simple seperation of concerns tell us one object should not be responsible for both being the catalog and loading external resources.

This single method realy makes people struggle with using this in things like intercepters, config functions etc.

**Comment by rubenv** [comment]
I agree, it's a bit unfortunate this got added (though it does keep things simple for many people).
Can't be removed without breaking API either.

**Comment by frederikprijck** [comment]
I understand, however apart from totally liking gettext I think keeping this inside the service will pull people away from it (or atleast those who are aware of the issue) while I rly think this shit's realy good man.

All it does is make it simple for people who are less aware of what's supposed to be happening. I don't want to be dependend on $http when all I want to do is ask for the current language, that's totally unacceptable, neither should I be dependend on $http to control a translations catalog, but this latter is not the biggest issue.

My problem is this:
- I have a LanguageService, dependend on gettextCatalog
- I have a LanguageInterceptor to make sure the language get's added to the http headers, which is dependend on LanguageService

What happens is the following:
- Angular registers gettextCatalog and resolves the $http dependency.
- While resolving $http, angular resolves all registered interceptors
- As my interceptor requires languageService (to get the current language), the languageService is resolved by angular.
- As the languageService is dependend on gettextCatalog, angular resolved gettextCatalog and we are back at the start (circular references).

Injecting $injector and using the Service Locator pattern doesn't work (or I'm missing something).

Currently, I do not see a proper workaround besides storing the current language on 2 services (one to simply set/get the language from memory, the other one is the gettextCatalog).
As I'm using a LanguageService, both services are controller by this languageService. So the downside isn't very big, but it's pitty that we have the need to create such a workaround because gettext doesn't support the most basic feature of a translation library: getting the current language without being dependend on irrelevant things.

**Comment by zirand** [comment]
I need to setup 'Accept-Language' header. however I must do it via  httpProvider.interceptors. And i get lang frоm appConfig. And You are not belive but I am setup gettextCatalog.setCurrentLanguage(lang) in appConfig [sarcasm]!!! And Miracle!!!   
[$injector:cdep] Circular dependency found: gettextCatalog < - appConfig < - $http < - gettextCatalog

:-(

**Comment by MigFerreira** [comment]
Having the same problem with the $httpInterceptor... would really love this to be solved as i cant seem to get around it. Maybe best option is changing to another translation solution.

**Comment by frederikprijck** [comment]
@zirand and @MigFerreira  have you tried my workaround, mentioned above ? I got it working ...

**Comment by MigFerreira** [comment]
@zirand, @frederikprijck and @rubenv  I actually removed $http direct dependency from gettextCatalog by changing loadRemote to:

```javascript
loadRemote: function (url) {
            return $injector.get('$http')({
                method: 'GET',
                url: url,
                cache: catalog.cache
            }).then(function (response) {
                var data = response.data;
                for (var lang in data) {
                    catalog.setStrings(lang, data[lang]);
                }
                return response;
            });
}
```
Did this 10 min ago, and seems to be working but need to do some more tests.

**Comment by frederikprijck** [comment]
Well IIRC I tried this aswell and it didn't work. But maybe you're having better luck. Please to keep us informed on your tests !

**Comment by MigFerreira** [comment]
@frederikprijck @rubenv @zirand, Ok tested all my use cases and my solution works, don't know if its a 100% solution for all possible use cases. 

My problematic use cases:
    languageService loading json remotely when uses changes language in preferences.
    httpInterceptor translating api errors based on the response codes matching a service of errorConstants.
The errorConstants service was using gettextCatalog and was generating the circular dependency. Now works fine.

I really hate changing external lib source, when i have more time i will try your solution @frederikprijck.

**Comment by frederikprijck** [comment]
I'm nowhere saying that my solution is better then urs. I highly doubt it. Your solution requires less effort, however it does require modification of source code which I don't do unless it's being accepted as PR...
